imdsclient: better retries with tokio retry and timeout

jpculp opened the pull request:
**Description of changes:**

To mitigate things like closed connections, IMDSv2 token and data
fetches will continue to retry until success or until the timeout has
been reached.

The default timeout is 300s to match the ifup timeout in wicked, but can
be set manually with `.with_timeout` when building your own ImdsClient.

A wait_for_imds function was also added if you simply want to verify
IMDS is available before proceeding.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

jpculp committed:
imdsclient: better retries with tokio retry and timeout

To mitigate things like closed connections, IMDSv2 token and data
fetches will continue to retry until success or until the timeout has
been reached.

The default timeout is 300s to match the ifup timeout in wicked, but can
be set manually using `.with_timeout` when creating a new ImdsClient.

webern commented on sources/imdsclient/src/lib.rs:
```diff
@@ -16,26 +16,34 @@ The result is returned as a `String` _(ex. m5.large)_.
     imds_base_uri: String,
-    session_token: String,
+    retry_timeout: Duration,
+    session_token: Arc<RwLock<String>>,
```
`Arc` may not be needed since the ImdsClient is not `Clone`.

jpculp replied:
I think you are correct. Everything seems to work with just RwLock<String>.

webern commented on sources/imdsclient/src/lib.rs:
```diff
@@ -221,91 +240,91 @@ impl ImdsClient {
+        timeout(
+            self.retry_timeout,
+            Retry::spawn(retry_strategy(), || async {
+                let session_token = {
```
I think this async block should be a function defined elsewhere and called here (for readability).

webern commented on sources/imdsclient/src/lib.rs:
```diff
@@ -221,91 +240,91 @@ impl ImdsClient {
-                }
+        timeout(
+            self.retry_timeout,
+            Retry::spawn(retry_strategy(), || async {
```
I haven't used this before. It looks like it will retry for any error? Some errors are not worth retrying, right?

cbgbt replied:
`RetryIf` exists in cases where we want to conditionally retry, but the only one I can think of is 404, which is handled by returning `None` in this case.

jpculp commented:
- Removed `wait_for_imds` and other pieces associated with it.
- Removed Arc.
- Added a token accessor `read_token` to simplify reading the token.
- Cleaned up the error syntax in some of the `fetch_imds` paths.

jpculp committed:
imdsclient: move token fetch out ImdsClient::new()

This moves the token fetch to just before the IMDS data fetch so
that building a new ImdsClient no longer needs an await.

For better readability refresh_token is now write_token and a read_token
accessor has been added.

jpculp commented:
- `read_token` accessor was moved to a second commit and no longer uses `.unwrap()`.
- `session_token` is now an `RwLock<Option<String>>`.
- `ImdsClient::new()` is no longer asynchronous.
- `.await` removed from ebc, pluto, and shibaken.

jpculp commented:
This changes the logic to read the client token once, checking if the value is none, and if so writing a new token. This allows for tokens to be shared between calls (like before), but slows down the very first call by 500ms since it triggers a retry.

webern reviewed (commented):
lgtm (quick readthrough)
